torch/csrc/jit/script/module.h

```diff
@@ -341,7 +341,7 @@ struct slot_iterator_impl {
     }
     // the last traversal action advanced beyond the number of slots in the
     // module so continue the iteration in the parent.
-    if (top().i_ >= int64_t(top().module_.num_slots())) {
+    if (top().i_ >= int64_t(top().module_._ivalue()->type()->numAttributes())) {
```

Comment:
@zdevito @suo when we remove attributes for the ClassType that's shared by multiple module instances, there is an inconsistent state where we have removed these attributes in ClassType, but not in the module instance, that's why this change is needed here. If the module is in a consistent state `module_.num_slots()` will always be equal to `module_._ivalue()->type()->numAttributes()` and this change will have no effect. What do you think?
I think we can either do this or create a new Policy for inconsistent modules.

Reply:
I think it is a big no-no to have the module object state inconsistent with the type. This change proposes the hide the fact that they are inconsistent by switching around the source of truth for a public API, which seems like it will be confusing later.

Is there a way to do this where we never make the modules inconsistent? Two suggestions I see (granted I don't know that much about what this pass is doing so perhaps these are not useful):
1. Don't delete attributes, just construct a new class type with the attributes deleted. (are we already doing this?)
2. When we remove the attributes from the type, also remove them from the module object.

Reply:
The problem is the ClassType is shared by multiple module instances and we want to delete some attribute from the ClassType(and remove slots from the object). (responding to 2.)Suppose there are two module instances share the ClassType, we can safely delete the attribute and also remove the slot from the first module instance, but after that the second module instance will be in a inconsistent state where it has more slots than attributes.
(responding to 1.) If we create a new ClassType, we need to swap out the original ClassType with the new one, and I think that is unsafe as well, and we'll need to do type remapping which also requires more analysis on where the module instance is used.

This is unsafe in general, but in the context of quantization, we know that we inserted the observers ourselves in the insert_observers pass and here we are removing them so I think it is OK to temporary having inconsistent modules in this case.

Reply:
Okay, as far as the change to the core, I am okay with this. The contract is still that modules are consistent with their types, and in that case this change is a no-op.

I am a little skeptical about this inconsistency you are describing and how easy it will be to maintain for the quant passes, but I guess that's fundamentally up to you. If you'd like we can chat a bit about it; in particular I don't understand how your proposal doesn't error the second time you try `type()->unsafeRemoveAttribute()`—didn't you already remove it for the first module?

Reply:
Yes, second time I'm replaying the removal of slots by index :)

Reply:
I want to talk more about the details for documentation purpose. 
## Inconsistency
### Type 1(`num_slots` < `numAttributes`)
I think the current core has one type of inconsistency which is `num_slots` can be less then `numAttributes` in the type(https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/module.h#L131-L138), in Module::register_attribute, we first `type()->addOrCheckAttribute` and then do `_ivalue()->setAttr(...)` and `setAttr` which calls `setSlot` will automatically create a new slot if the slot index is greater or equal to the size of the slot list, this means after `type()->addOrCheckAttribute` we have more attributes than slots.
### Type 2(`num_slots` > `numAttributes`)
In this PR we introduce another type of inconsistency, which is `numAttributes` is less than `num_slots`, when we remove attribute of a shared ClassType and the slot is not yet removed for some module instances.

## ModulePolicy
Previously we use `num_slots` as the ground truth, and Type 1 inconsistency is the only inconsistency we see, the inconsistency won't be visible, even we iterate through a half constructed module. However, the new type of inconsistency will be visible when we iterate through using the old ModulePolicy, since we have `num_slots` > `numAttributes`.

That's why I changed the ModulePolicy to use `numAttributes` as ground truth, this means if we iterate through a module that has the Type 1 inconsistency, we will see error, but Type 2 inconsistency will not be visible when we iterate use the new ModulePolicy.

I'm assuming Type 1 inconsistency is only there during module construction and after construction we should either have Type 2 inconsistent(as it appears here) objects or consistent objects.

Reply:
I also squinted a bit about that new inconsistency, but after reading the code carefully I think it should be fine. My ask, however, would be to try to make this code crystal clear - it should be obvious where the inconsistency starts, where it ends and why we never misuse inconsistent objects (e.g. try to double delete).

Reply:
It's all constraint in the insert_quant_dequant pass